src/audio/drc/drc.c

```diff
@@ -0,0 +1,605 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2020 Google LLC. All rights reserved.
+//
+// Author: Pin-chih Lin <johnylin@google.com>
+
+#include <sof/audio/component.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/drc/drc.h>
+#include <sof/audio/drc/drc_math.h>
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/ipc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/math/decibels.h>
+#include <sof/platform.h>
+#include <sof/string.h>
+#include <sof/ut.h>
+#include <sof/trace/trace.h>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/eq.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static const struct comp_driver comp_drc;
+
+/* TODO: get the real UUID */
+/* 5150c0e6-27f9-4ec8-8351-c705b642d12f */
+DECLARE_SOF_RT_UUID("drc", drc_uuid, 0x5150c0e6, 0x27f9, 0x4ec8,
+		    0x83, 0x51, 0xc7, 0x05, 0xb6, 0x42, 0xd1, 0x2f);
+
+DECLARE_TR_CTX(drc_tr, SOF_UUID(drc_uuid), LOG_LEVEL_INFO);
+
+static inline void drc_free_parameters(struct sof_drc_config **config)
+{
+	rfree(*config);
+	*config = NULL;
+}
+
+static inline void drc_reset_state(struct drc_state *state)
+{
+	int i;
+
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; ++i) {
+		rfree(state->pre_delay_buffers[i]);
+		state->pre_delay_buffers[i] = NULL;
+	}
+
+	state->detector_average = 0;
+	state->compressor_gain = Q_CONVERT_FLOAT(1.0f, 30);
+
+	state->last_pre_delay_frames = DRC_DEFAULT_PRE_DELAY_FRAMES;
+	state->pre_delay_read_index = 0;
+	state->pre_delay_write_index = DRC_DEFAULT_PRE_DELAY_FRAMES;
+
+	state->envelope_rate = 0;
+	state->scaled_desired_gain = 0;
+
+	state->processed = 0;
+
+	state->max_attack_compression_diff_db = INT32_MIN;
+}
+
+static inline int drc_init_pre_delay_buffers(struct drc_state *state,
+					     size_t sample_bytes,
+					     int channels)
+{
+	int i;
+
+	/* Allocate pre-delay (lookahead) buffers */
+	for (i = 0; i < channels; ++i) {
+		state->pre_delay_buffers[i] =
```

Comment:
Would it be better to allocate in a single rzalloc() the combined need of buffers? Since the pool is organized as 2^N sizes there typically remains unused but consumed RAM with non-2^N sizes.